include/pybind11/detail/function_record_pyobject.h

```diff
@@ -46,12 +46,27 @@ static PyMethodDef tp_methods_impl[]
        {nullptr, nullptr, 0, nullptr}};
 
 // Note that this name is versioned.
-constexpr char tp_name_impl[]
+constexpr char tp_qualname_impl[] = PYBIND11_INTERNAL_MODULE_NAME
+    "."
+    "pybind11_detail_function_record_" PYBIND11_DETAIL_FUNCTION_RECORD_ABI_ID
+    "_" PYBIND11_PLATFORM_ABI_ID;
+constexpr char tp_plainname_impl[]
```

Comment:
Did you consider adding
```
setattr((PyObject *) type, "__module__", str(PYBIND11_INTERNAL_MODULE_NAME));
```
in `get_function_record_PyTypeObject()`, instead of making this change here? That would seem more idiomatic and consistent within pybind11.

Reply:
Yes, I tried that but it doesn't work and gives:

`DeprecationWarning: builtin type pybind11_detail_function_record_v1_system_libstdcpp_gxx_abi_1xxx_use_cxx11_abi_1 has no __module__ attribute`

I don't really understand why.

Reply:
Hm ... I think it's worth spending time getting to the bottom of this. I might be able to drill down in a few hours.

**Wild guess:** Maybe we need to explicitly add `Py_TPFLAGS_HEAPTYPE`?

Reply:
Explicitly adding `Py_TPFLAGS_HEAPTYPE` does not help, I'm still seeing this:

```
<frozen importlib._bootstrap>:488: 321 warnings
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type pybind11_detail_function_record_v1_system_libstdcpp_gxx_abi_1xxx_use_cxx11_abi_1 has no __module__ attribute
```

I'll try to find out why.